wordpress-mobile/WordPress-Android
# reset uploading media to failed if the UploadService doesn't have them

Fixes #6455 

This PR does the following, under the conditions described in #6455:
Assuming you had some Post with media upload going on, or even a Post with failed media items in it, then you stop the app or turn the phone off/on.
- When you open the app again, there are no records of this having ever happened. It is only when you tap on the Post you were working on and open it in the Editor again now, that we're running through its content and check for any media item marked `UPLOADING`. These are all marked `FAILED` now, so they can be properly represented by the Editor with the failed / retry overlay. (*)
- Also, if we had any `UPLOADING` or `FAILED` items that for some reason already have a remote URL in their `src`  tags within the Post Content, these are unmarked as these are already valid URLs for the Post.
- There's still the case where you can see images or videos in the failed/retry state and tap on them to retry. This can happen when an image has been uploaded successfully, and so the `MediaStore` has the right information, but the `PostModel` has not yet been updated with the right information (that is, giving the `src` attribute the remote url value) when the crash / stop happens. In this case, the  user taps on the image, and the Retry callback will check whether the image already exist in the `MediaStore` and it has a remoteUrl, and just use it, recovering the image immediately). 
- **Note 1** (*): Given this is not a process we run for all posts given it could be a long running task, we rather do this only for each post as you try to open it in the editor. That's why there's no indication of previous media failed items on the Post list right after coming back from a crash/force stop. So, it is expected that you'll see your failed Post (the one you were working on before the crash) as a `Local draft` in the Posts list.
- **Note 2**: if you have several media items marked failed, and start uploading some of them (but not all), and go back, they will still be loading in the background, but the Post in the posts list will not show any progress, so giving more importance to failed media items that need action from the user (this was also the behavior before this PR).
If you start uploading them all (i.e. no failed items marked), we’ll be showing the progress as usual.

A better approach to this could be made, showing all failed Posts in the post list at once, recovering from any crash, once we[ move the `UploadStatus` handling to FluxC](https://github.com/wordpress-mobile/WordPress-Android/issues/4341#issuecomment-319413063). In the meantime, this PR tackles the problem described in #6455 which you cannot possibly recover from (blocking issue) without deleting all inconsistent media items and starting again.

To test:
1. start a new draft
2. include some media items
3. while these are uploading, stop the app (go to settings-> apps -> WordPress -> force stop)
4. open the app again
5. open the draft
6. observe failed media items show the failed/retry overlay
7.  tap on any failed item and observe the it either starts uploading again, or recovers its state gracefully if it's already found in the MediaStore
8. also tap on any non-failed item and observe the ImageSettings screen for such item is displayed, as expected.

cc @aforcier
